clash-prelude/src/Clash/Explicit/BlockRam/File.hs

```diff
@@ -205,6 +221,79 @@ blockRamFile = \clk gen sz file rd wrM ->
       (blockRamFile# clk gen sz file (fromEnum <$> rd) en (fromEnum <$> wr) din)
 {-# INLINE blockRamFile #-}
 
+-- | Convert data to the String contents of a memory file.
+--
+-- * __NB:__ Not synthesizable
+-- * The following document the several ways to instantiate components with
+-- files:
+--
+--     * "Clash.Prelude.BlockRam.File#usingramfiles"
+--     * "Clash.Prelude.ROM.File#usingromfiles"
+--     * "Clash.Explicit.BlockRam.File#usingramfiles"
+--     * "Clash.Explicit.ROM.File#usingromfiles"
+--
+-- * See "Clash.Sized.Fixed#creatingdatafiles" for more ideas on how to create
+-- your own data files.
+--
+-- = Example
+--
+-- The @Maybe@ datatype has don't care bits, where the actual value does not
+-- matter. But the bits need a defined value in the memory. Either 0 or 1 can be
+-- used, and both are valid representations of the data.
+--
+-- >>> let es = [ Nothing, Just (7 :: Unsigned 8), Just 8]
+-- >>> mapM_ (putStrLn . show . pack) es
+-- 0b0_...._....
+-- 0b1_0000_0111
+-- 0b1_0000_1000
+-- >>> putStr (memFile (Just 0) es)
+-- 000000000
+-- 100000111
+-- 100001000
+-- >>> putStr (memFile (Just 1) es)
+-- 011111111
+-- 100000111
+-- 100001000
+--
+memFile
+  :: forall a
+   . ( BitPack a
+     , HasCallStack)
+  => Maybe Bit
+  -- ^ Value to map don't care bits to. Nothing means throwing an error on
+  -- don't care bits.
+  -> [a]
+  -- ^ Values to convert.
+  -> String
+  -- ^ Contents of the memory file.
+memFile care = foldr (\e -> showsBV $ pack e) ""
+ where
+  showsBV :: BitVector (BitSize a) -> String -> String
+  showsBV (BV mask val) s =
+    if n == 0 then
+      '0' : '\n' : s
+    else
+      case care of
+        Just (Bit 0 0) -> go n (val .&. (mask `xor` fullMask)) ('\n' : s)
+        Just (Bit 0 1)  -> go n (val .|. mask) ('\n' : s)
```

Comment:
It's unlikely that `low` and `high` will change, but maybe it's better to do

```haskell
case care of
  Just b
    | b == low -> ...
    | b == high -> ...
```

Reply:
That'll error on don't care bits!

Reply:
I don't care about those :wink:

But more seriously, I thought they had a non-zero mask anyway? Or is that only undefined bits?

Reply:
s/bits/bit/ :-)

Reply:
I never saw your comment before, odd. I was s/-ing my own "undefined bits" comment. The problem is that `Eq Bit` correctly stumbles over undefined bits. After all, you want `lsb $(bLit ".") == 0` to throw an error instead of taking a wild stab at a boolean result.

Reply:
Ah yes, I see now. I guess you would need to use something like `Clash.Class.BitPack.isLike` to compare in a way that's friendly to don't care values

Reply:
The semantics of that here would mean that you want don't care bits to be mapped to don't care bits. And that is precisely the semantics of the current code. `Just $ lsb $(bLit ".")` is handled the same as `Nothing` by the `_` pattern match.

Reply:
User code should probably not peek under the hood of `Bit`, but we can do that, it's the same body of code as where it is defined.